File: tasks/rowofdices/solutions/ajreme-ok.py
N, prime = int(input()), [-1]*20000

for i in range(2, 20000):
	if prime[i] == -1:
		prime[i] = i
		for j in range(i*i, 20000, i):
			prime[j] = 0
prime = [*filter(lambda x: prime[x]==x, prime)]

def solve(cur=1, ptr=0, need4=0):
	ret = 1-need4
	for i in range(ptr, len(prime)):
		if cur*prime[i]**(4 if need4 else 6) > N:
			break
		for j in range(6, 60, 6):
			if cur*prime[i]**j > N:
				break
			ret += solve(cur*prime[i]**j, i+1, need4)
		for j in range(4, 60, 6):
			if cur*prime[i]**j > N:
				break
			ret += solve(cur*prime[i]**j, i+1, 1-need4)
	return ret

print(solve())

File: tasks/rowofdices/solutions/ajreme-ok.cpp
#include <bits/stdc++.h>

using namespace std;

long long n;
bool isprime[20000] = {};
vector<int> prime;

void sieve() {
    for (int i = 2; i < 20000; i++) {
        if (!isprime[i]) {
            prime.push_back(i);
            for (int j = i*i; j < 20000; j += i) {
                isprime[j] = 1;
            }
        }
    }
}

long long powll(long long a, long long b) {
    if (b == 0) {
        return 1;
    }
    long long temp = powll(a, b/2);
    return temp * temp * (b % 2 != 0 ? a : 1);
}

bool cant_mul(long long cur, int p, int s) {
    return 1.0l*cur*pow(1.0l*prime[p], s) > n;
}

int solve(long long cur = 1, int ptr = 0, int need4 = 0) {
    int ret = 1 - need4;
    for (int i = ptr; i < prime.size(); i++) {
        if (cant_mul(cur, i, (need4 == 1 ? 4 : 6))) {
            break;
        }
        for (int j = 6; j < 60; j += 6) {
            if (cant_mul(cur, i, j)) {
                break;
            }
            ret += solve(cur*powll(prime[i], j), i+1, need4);
        }
        for (int j = 4; j < 60; j += 6) {
            if (cant_mul(cur, i, j)) {
                break;
            }
            ret += solve(cur*powll(prime[i], j), i+1, 1 - need4);
        }
    }
    return ret;
}

int main() {
    cin >> n;
    sieve();
    cout << solve();
}
